Support distributed encoding

shssoichiro opened the pull request:
Adds rav1e-worker, which acts as a server on a remote encoding machine and listens to connections from clients. The client, part of rav1e-by-gop, negotiates with the server, the server notifies the client as it has slots available, and the client sends one segment at a time to the server. The server can have multiple worker threads, and a client can use as many threads as the server will allow. A server can also serve multiple clients at once, which it does via a FIFO queue. The queue only holds slot requests, so the client does not send frames until the server signals a slot is ready. The client will send to the first worker that signals a slot is available. This keeps memory usage down on both the client and server, and streamlines the threading process so the end of an encode is not stalled waiting for a specific remote worker, if there are other workers or local threads available.

Communication is done using websockets, and is insecure by default, but can be configured to use TLS. Unfortunately this does not currently work with self-signed certs. Client-server authentication is done using a simple password system; a password is configured on the server, and then the client must be configured to use the same password for this server.

The frame sending process is quite upload-heavy. The input frames are sent xz-compressed to try and alleviate this, however results will still be best if workers are on a LAN or other connection with a fast upload. I found that even with a 10Mbps upload from my home connection, the remote workers still were more than efficient enough to be worth using.

Closes #10

lu-zero commented:
Can we split this PR in 3?
- Refactor (it should land immediately)
- Client/Server structure (should enjoy some patch juggling)
- Refinements (it might require some discussion and I might have few more suggestions)
  - raw content compression (e.g. xz or zstd?)
  - auth

shssoichiro commented:
@lu-zero I'm opting to reuse this PR for discussion; I've rebased off master. I think it's fine to have the discussion on compression and auth here as well. If we need to split something out into another PR e.g. more complex auth support, that's reasonable, but at the moment the compression and auth implementations are fairly small code-wise.

shssoichiro commented:
Seems to be some big memory leaking going on... so, going to investigate.

lu-zero commented on rav1e-by-gop/src/bin/encode.rs:
```diff
@@ -138,19 +185,96 @@ pub fn perform_encode_inner<T: Pixel, R: 'static + Read + Send>(
+        );
     });
 
+    for worker in remote_slots.lock().unwrap().iter() {
```
This block could be a stand alone function maybe?

lu-zero commented on rav1e-by-gop/src/bin/encode.rs:
```diff
@@ -166,17 +290,53 @@ pub fn perform_encode_inner<T: Pixel, R: 'static + Read + Send>(
     let mut num_segments = 0;
     let encode_opts = EncodeOptions::from(opts);
     let output_file = opts.output.to_owned();
+
```
this as well.

lu-zero commented on rav1e-worker/src/worker.rs:
```diff
@@ -0,0 +1,230 @@
+            ));
+
+            // Spawn the encoder thread for this connection within the encoder thread pool
+            thread_pool.execute(move || {
```
This should be a stand alone function as well.

lu-zero commented on rav1e-worker/src/worker.rs:
```diff
@@ -0,0 +1,230 @@
+    let connection_id = input.connection_id;
+    let frames = input.frames.into_iter().map(Arc::new).collect::<Vec<_>>();
+    if !frames.is_empty() {
+        let mut cfg = Config {
```
This could be a stand alone function as well.

shssoichiro commented:
I'm only noticing the high memory usage when dealing with 1080p+ videos and a slow upload to the remote server (slow = 10mbps). It may be the leak is still occurring in other situations but is not as noticeable. I haven't been able to track down what's causing it, but my suspicions are either xz or the websocket connection or underlying network connection itself is causing it.

lu-zero commented:
I guess frames end up piling up somewhere.

shssoichiro commented:
I'm going to do some more testing on the memory usage and see if I can pin it down. In the meantime, @lu-zero you mentioned you had some opinions on compression and authorization. Would you mind sharing those?

lu-zero commented:
Sure:)
- Regarding compression probably zstd or lz4 might have a better tradeoff compared to xz.
- For auth I hadn't read in depth what's in place. 
  - I guess some deploys would enjoy having no-auth while other might prefer having other auth systems.
- Transport-wise tls/plain is good.

shssoichiro commented:
Fixed a couple of bugs, and switched over to zstd. Zstd does seem to provide a better speed-to-compression ratio. I can't even seem to replicate the memory issue consistently, so for all I know it might be KDE doing some weird thumbnailing stuff or something.

The current auth implementation uses a simple password shared between the server and client. The client sends the password as a header when it makes the connection, and the server rejects the connection if the password doesn't match. A password is currently required--I really don't like having no-password as a default, so maybe there could be an explicit no-password option if that's a desire for some users.

lu-zero commented:
Since the tls transport could mutually identify the endpoints a password would be unnecessary :) but those are small details I guess.

lu-zero commented:
Since the tls transport could mutually identify the endpoints a password would be unnecessary :) but those are small details I guess.

shssoichiro commented:
In this case I'm not sure how TLS would be used for what I'm trying to accomplish with the password. Since TLS automatically assumes the client is trustworthy if it has the required ciphers installed. The intent behind the password is, for example, if I have a worker running on a server that's internet-facing, I want to prevent random users who may find out that the port is open from loading up my worker with tasks and CPU usage. (This is such obscure software at the moment that it's unlikely to be targeted by such malicious users, but even in a case like this, I try to avoid security through obscurity.)

tleydxdy commented:
TLS can have client certificate

shssoichiro commented:
After some consideration, I thought about using the TLS client cert, but I'd prefer to adopt an approach that also works for unencrypted connections and is secured by default. In reality, I'd love to make this TLS-only, but given that I was unable to get it to work with self-signed certs, that's not going to be feasible for people who want to use this over a LAN. So I think that requiring a password by default is a reasonable compromise.
